File: Grind75_1/week3/01Matrix.py
"""
https://leetcode.com/problems/01-matrix/description/
"""

class Solution:

    # Direction vectors
    directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]

    lenOfRows = 0
    lenOfCols = 0

    # Function to check if a cell
    # is be visited or not
    def isValid(self, vis, row, col):
        
        # If cell lies out of bounds
        if (row < 0 or col < 0 or row >= self.lenOfRows or col >= self.lenOfCols):
            return False
    
        # If cell is already visited
        if ((row, col) in vis):
            return False
    
        # Otherwise
        return True

    # https://www.geeksforgeeks.org/breadth-first-traversal-bfs-on-a-2d-array/
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:

        self.lenOfRows = len(mat)
        self.lenOfCols = len(mat[0])

        # Stores indices of the matrix cells
        queue = deque()

        # Declare the visited set
        visited = set()

        # mark all 0 elements as visited
        for x in range(self.lenOfRows):
            for y in range(self.lenOfCols):
                if (mat[x][y] == 0):
                    # marked
                    visited.add((x, y))
                    # added to queue to later check its neighbors
                    queue.append((x, y))

        # BFS
        while (queue):
            
            x, y = queue.popleft()
            
            # check neighbors
            for direct in self.directions:
                adjx = x + direct[0]
                adjy = y + direct[1]
                
                if (self.isValid(visited, adjx, adjy)):
                    # Bottom-Up Approach (Tabulation)
                    mat[adjx][adjy] = mat[x][y] + 1
                    visited.add((adjx, adjy))
                    queue.append((adjx, adjy))

        return mat

File: Grind75/Graph/01-matrix.js
// https://leetcode.com/problems/01-matrix/description/

/**
BFS

create visited Array
create queue

for all 0 cells
    add to visited
    enqueue to queue to search its neighbors

while qu not empty
    [r,c] = popFront

    for each direction
        if out of bounds or visited
            continue

        // if here, then the not visited is a 1 so the distance to a 0 is:
        mat[dr][dc] = mat[r][c] + 1     // if mat[r][c] is 0 then it is: 0 + 1. if mat[r][c] is a 1, then it has already been evaluated (since in queue) and holds the closest 0 dist, therefore + 1 will give this cell the closest
        visited[dr][dc] = true
        qu.pushBack([dr, dc])

return mat

- Time: O(m * n)
- Space: O(m * n)
 */

/**
 * @param {number[][]} mat
 * @return {number[][]}
 */
var updateMatrix = function(mat) {
    const rows = mat.length
    const cols = mat[0].length
    const directions = [[-1, 0], [0, 1], [1, 0], [0, -1]]
    const res = Array.from(new Array(rows), (e) => new Array(cols).fill(0))
    const visited = new Array(rows).fill().map((e) => new Array(cols).fill(false))
    const qu = new Deque()

    for (let r = 0; r < rows; r ++) {
        for (let c = 0; c < cols; c ++) {
            if (mat[r][c] === 0) {
                visited[r][c] = true
                qu.pushBack([r, c])
            }
        }
    }

    while (qu.size() > 0) {
        const [r, c] = qu.popFront()
        for (let [dr, dc] of directions) {
            const nr = r + dr
            const nc = c + dc
            if (nr < 0 || nr >= rows || nc < 0 || nc >= cols || visited[nr][nc] === true) {
                continue
            }

            mat[nr][nc] = mat[r][c] + 1
            visited[nr][nc] = true
            qu.pushBack([nr, nc])
        }
    }

    return mat
};